File: gatling-http/src/main/scala/io/gatling/http/protocol/HttpProtocolBuilder.scala
/*
 * Copyright 2011-2021 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.http.protocol

import java.net.{ Inet4Address, InetAddress, InetSocketAddress }
import java.util.regex.Pattern
import javax.net.ssl.KeyManagerFactory

import scala.jdk.CollectionConverters._

import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.filter.{ BlackList, Filters, WhiteList }
import io.gatling.core.session._
import io.gatling.core.session.el.El
import io.gatling.http.{ MissingNettyHttpHeaderNames, ResponseTransformer }
import io.gatling.http.check.HttpCheck
import io.gatling.http.client.SignatureCalculator
import io.gatling.http.client.realm.Realm
import io.gatling.http.client.uri.Uri
import io.gatling.http.fetch.InferredResourceNaming
import io.gatling.http.request.builder.RequestBuilder
import io.gatling.http.util.{ HttpHelper, InetAddresses }

import com.softwaremill.quicklens._
import io.netty.handler.codec.http.HttpHeaderNames
import io.netty.handler.ssl.SslProvider

object HttpProtocolBuilder {

  implicit def toHttpProtocol(builder: HttpProtocolBuilder): HttpProtocol = builder.build

  def apply(configuration: GatlingConfiguration): HttpProtocolBuilder =
    HttpProtocolBuilder(HttpProtocol(configuration), configuration.ssl.useOpenSsl)
}

final case class HttpProtocolBuilder(protocol: HttpProtocol, useOpenSsl: Boolean) {

  def baseUrl(url: String): HttpProtocolBuilder = baseUrls(List(url))
  def baseUrls(urls: String*): HttpProtocolBuilder = baseUrls(urls.toList)
  def baseUrls(urls: List[String]): HttpProtocolBuilder = this.modify(_.protocol.baseUrls).setTo(urls)
  def warmUp(url: String): HttpProtocolBuilder = this.modify(_.protocol.warmUpUrl).setTo(Some(url))
  def disableWarmUp: HttpProtocolBuilder = this.modify(_.protocol.warmUpUrl).setTo(None)

  // enginePart
  def shareConnections: HttpProtocolBuilder = disableAutoReferer.modify(_.protocol.enginePart.shareConnections).setTo(true)
  def virtualHost(virtualHost: Expression[String]): HttpProtocolBuilder = this.modify(_.protocol.enginePart.virtualHost).setTo(Some(virtualHost))
  def localAddress(address: String): HttpProtocolBuilder = localAddresses(address :: Nil)
  def localAddresses(addresses: String*): HttpProtocolBuilder = localAddresses(addresses.toList)
  def localAddresses(addresses: List[String]): HttpProtocolBuilder = {
    val (ipV4Addresses, ipV6Addresses) = addresses.map(InetAddress.getByName).partition(_.isInstanceOf[Inet4Address])
    localAddresses(ipV4Addresses, ipV6Addresses)
  }
  def useAllLocalAddresses: HttpProtocolBuilder = useAllLocalAddressesMatching()
  def useAllLocalAddressesMatching(patterns: String*): HttpProtocolBuilder = {
    val compiledPatterns = patterns.map(Pattern.compile)

    def filter(addresses: List[InetAddress]): List[InetAddress] =
      addresses.filter { address =>
        val hostAddress = address.getHostAddress
        compiledPatterns.exists(_.matcher(hostAddress).matches)
      }

    localAddresses(filter(InetAddresses.AllIpV4LocalAddresses), filter(InetAddresses.AllIpV6LocalAddresses))
  }

  private def localAddresses(ipV4Addresses: List[InetAddress], ipV6Addresses: List[InetAddress]): HttpProtocolBuilder =
    this
      .modify(_.protocol.enginePart.localIpV4Addresses)
      .setTo(ipV4Addresses)
      .modify(_.protocol.enginePart.localIpV6Addresses)
      .setTo(ipV6Addresses)

  def maxConnectionsPerHostLikeFirefoxOld: HttpProtocolBuilder = maxConnectionsPerHost(2)
  def maxConnectionsPerHostLikeFirefox: HttpProtocolBuilder = maxConnectionsPerHost(6)
  def maxConnectionsPerHostLikeOperaOld: HttpProtocolBuilder = maxConnectionsPerHost(4)
  def maxConnectionsPerHostLikeOpera: HttpProtocolBuilder = maxConnectionsPerHost(6)
  def maxConnectionsPerHostLikeSafariOld: HttpProtocolBuilder = maxConnectionsPerHost(4)
  def maxConnectionsPerHostLikeSafari: HttpProtocolBuilder = maxConnectionsPerHost(6)
  def maxConnectionsPerHostLikeIE7: HttpProtocolBuilder = maxConnectionsPerHost(2)
  def maxConnectionsPerHostLikeIE8: HttpProtocolBuilder = maxConnectionsPerHost(6)
  def maxConnectionsPerHostLikeIE10: HttpProtocolBuilder = maxConnectionsPerHost(8)
  def maxConnectionsPerHostLikeChrome: HttpProtocolBuilder = maxConnectionsPerHost(6)
  def maxConnectionsPerHost(max: Int): HttpProtocolBuilder = this.modify(_.protocol.enginePart.maxConnectionsPerHost).setTo(max)
  def perUserKeyManagerFactory(f: Long => KeyManagerFactory): HttpProtocolBuilder = this.modify(_.protocol.enginePart.perUserKeyManagerFactory).setTo(Some(f))

  // requestPart
  def disableAutoReferer: HttpProtocolBuilder = this.modify(_.protocol.requestPart.autoReferer).setTo(false)
  def disableCaching: HttpProtocolBuilder = this.modify(_.protocol.requestPart.cache).setTo(false)
  def header(name: CharSequence, value: Expression[String]): HttpProtocolBuilder = this.modify(_.protocol.requestPart.headers).using(_ + (name -> value))
  def headers(headers: Map[_ <: CharSequence, String]): HttpProtocolBuilder =
    this.modify(_.protocol.requestPart.headers).using(_ ++ headers.view.mapValues(_.el[String]))
  def acceptHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.ACCEPT, value)
  def acceptCharsetHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.ACCEPT_CHARSET, value)
  def acceptEncodingHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.ACCEPT_ENCODING, value)
  def acceptLanguageHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.ACCEPT_LANGUAGE, value)
  def authorizationHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.AUTHORIZATION, value)
  def connectionHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.CONNECTION, value)
  def contentTypeHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.CONTENT_TYPE, value)
  def doNotTrackHeader(value: Expression[String]): HttpProtocolBuilder = header(MissingNettyHttpHeaderNames.DNT, value)
  def userAgentHeader(value: Expression[String]): HttpProtocolBuilder = header(HttpHeaderNames.USER_AGENT, value)
  def upgradeInsecureRequestsHeader(value: Expression[String]): HttpProtocolBuilder = header(MissingNettyHttpHeaderNames.UpgradeInsecureRequests, value)
  def basicAuth(username: Expression[String], password: Expression[String]): HttpProtocolBuilder = authRealm(HttpHelper.buildBasicAuthRealm(username, password))
  def digestAuth(username: Expression[String], password: Expression[String]): HttpProtocolBuilder =
    authRealm(HttpHelper.buildDigestAuthRealm(username, password))
  def authRealm(realm: Expression[Realm]): HttpProtocolBuilder = this.modify(_.protocol.requestPart.realm).setTo(Some(realm))
  def silentResources: HttpProtocolBuilder = this.modify(_.protocol.requestPart.silentResources).setTo(true)
  def silentUri(regex: String): HttpProtocolBuilder = this.modify(_.protocol.requestPart.silentUri).setTo(Some(regex.r.pattern))
  def disableUrlEncoding: HttpProtocolBuilder = this.modify(_.protocol.requestPart.disableUrlEncoding).setTo(true)
  def sign(calculator: Expression[SignatureCalculator]): HttpProtocolBuilder = this.modify(_.protocol.requestPart.signatureCalculator).setTo(Some(calculator))
  def signWithOAuth1(
      consumerKey: Expression[String],
      clientSharedSecret: Expression[String],
      token: Expression[String],
      tokenSecret: Expression[String]
  ): HttpProtocolBuilder =
    sign(RequestBuilder.oauth1SignatureCalculator(consumerKey, clientSharedSecret, token, tokenSecret))
  def enableHttp2: HttpProtocolBuilder =
    if (SslProvider.isAlpnSupported(if (useOpenSsl) SslProvider.OPENSSL_REFCNT else SslProvider.JDK)) {
      this.modify(_.protocol.enginePart.enableHttp2).setTo(true)
    } else {
      throw new UnsupportedOperationException("You can't use HTTP/2 if OpenSSL is not available and Java version < 11")
    }

  def http2PriorKnowledge(remotes: Map[String, Boolean]): HttpProtocolBuilder =
    this
      .modify(_.protocol.enginePart.http2PriorKnowledge)
      .setTo(remotes.map { case (address, isHttp2) =>
        val remote = address.split(':') match {
          case Array(hostname, port) => new Remote(hostname, port.toInt)
          case Array(hostname)       => new Remote(hostname, 443)
          case _                     => throw new IllegalArgumentException("Invalid address for HTTP/2 prior knowledge: " + address)
        }
        remote -> isHttp2
      })

  // responsePart
  def disableFollowRedirect: HttpProtocolBuilder = this.modify(_.protocol.responsePart.followRedirect).setTo(false)
  def maxRedirects(max: Int): HttpProtocolBuilder = this.modify(_.protocol.responsePart.maxRedirects).setTo(max)
  def strict302Handling: HttpProtocolBuilder = this.modify(_.protocol.responsePart.strict302Handling).setTo(true)
  def transformResponse(responseTransformer: ResponseTransformer): HttpProtocolBuilder =
    this.modify(_.protocol.responsePart.responseTransformer).setTo(Some(responseTransformer))
  def check(checks: HttpCheck*): HttpProtocolBuilder = this.modify(_.protocol.responsePart.checks).using(_ ::: checks.toList)
  def inferHtmlResources(): HttpProtocolBuilder = inferHtmlResources(None)
  def inferHtmlResources(white: WhiteList): HttpProtocolBuilder = inferHtmlResources(Some(new Filters(white, BlackList.Empty)))
  def inferHtmlResources(white: WhiteList, black: BlackList): HttpProtocolBuilder = inferHtmlResources(Some(new Filters(white, black)))
  def inferHtmlResources(black: BlackList): HttpProtocolBuilder = inferHtmlResources(Some(new Filters(black, WhiteList.Empty)))
  def inferHtmlResources(black: BlackList, white: WhiteList): HttpProtocolBuilder = inferHtmlResources(Some(new Filters(black, white)))
  private def inferHtmlResources(filters: Option[Filters]): HttpProtocolBuilder =
    this
      .modify(_.protocol.responsePart.inferHtmlResources)
      .setTo(true)
      .modify(_.protocol.responsePart.htmlResourcesInferringFilters)
      .setTo(filters)
  def nameInferredHtmlResourcesAfterUrlTail: HttpProtocolBuilder = nameInferredHtmlResources(InferredResourceNaming.UrlTailInferredResourceNaming)
  def nameInferredHtmlResourcesAfterAbsoluteUrl: HttpProtocolBuilder = nameInferredHtmlResources(InferredResourceNaming.AbsoluteUrlInferredResourceNaming)
  def nameInferredHtmlResourcesAfterRelativeUrl: HttpProtocolBuilder = nameInferredHtmlResources(InferredResourceNaming.RelativeUrlInferredResourceNaming)
  def nameInferredHtmlResourcesAfterPath: HttpProtocolBuilder = nameInferredHtmlResources(InferredResourceNaming.PathInferredResourceNaming)
  def nameInferredHtmlResourcesAfterLastPathElement: HttpProtocolBuilder =
    nameInferredHtmlResources(InferredResourceNaming.LastPathElementInferredResourceNaming)
  def nameInferredHtmlResources(f: Uri => String): HttpProtocolBuilder = this.modify(_.protocol.responsePart.inferredHtmlResourcesNaming).setTo(f)

  // wsPart
  def wsBaseUrl(url: String): HttpProtocolBuilder = wsBaseUrls(List(url))
  def wsBaseUrls(urls: String*): HttpProtocolBuilder = wsBaseUrls(urls.toList)
  def wsBaseUrls(urls: List[String]): HttpProtocolBuilder = this.modify(_.protocol.wsPart.wsBaseUrls).setTo(urls)
  def wsReconnect: HttpProtocolBuilder = wsMaxReconnects(Int.MaxValue)
  def wsMaxReconnects(max: Int): HttpProtocolBuilder = this.modify(_.protocol.wsPart.maxReconnects).setTo(max)

  // proxyPart
  def noProxyFor(hosts: String*): HttpProtocolBuilder = this.modify(_.protocol.proxyPart.proxyExceptions).setTo(hosts)
  def proxy(httpProxy: Proxy): HttpProtocolBuilder = this.modify(_.protocol.proxyPart.proxy).setTo(Some(httpProxy.proxyServer))

  // dnsPart
  def asyncNameResolution(dnsServers: String*): HttpProtocolBuilder =
    asyncNameResolution(dnsServers.map { dnsServer =>
      dnsServer.split(':') match {
        case Array(hostname, port) => new InetSocketAddress(hostname, port.toInt)
        case Array(hostname)       => new InetSocketAddress(hostname, 53)
        case _                     => throw new IllegalArgumentException("Invalid dnsServer: " + dnsServer)
      }
    }.toArray)
  def asyncNameResolution(dnsServers: Array[InetSocketAddress]): HttpProtocolBuilder =
    this.modify(_.protocol.dnsPart.dnsNameResolution).setTo(AsyncDnsNameResolution(dnsServers))
  def hostNameAliases(aliases: Map[String, List[String]]): HttpProtocolBuilder = {
    val aliasesToInetAddresses = aliases.map { case (hostname, ips) =>
      hostname -> ips.map(ip => InetAddress.getByAddress(hostname, InetAddress.getByName(ip).getAddress)).asJava
    }
    this.modify(_.protocol.dnsPart.hostNameAliases).setTo(aliasesToInetAddresses)
  }
  def perUserNameResolution: HttpProtocolBuilder =
    this.modify(_.protocol.dnsPart.perUserNameResolution).setTo(true)

  def build: HttpProtocol = protocol
}

File: gatling-http-java/src/main/java/io/gatling/http/javaapi/HttpProtocolBuilder.java
/*
 * Copyright 2011-2021 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.http.javaapi;

import io.gatling.commons.validation.Validation;
import io.gatling.core.javaapi.CheckBuilder;
import io.gatling.core.javaapi.Filter;
import io.gatling.core.javaapi.ProtocolBuilder;
import io.gatling.core.javaapi.Session;
import io.gatling.http.client.SignatureCalculator;
import io.gatling.http.client.uri.Uri;
import io.gatling.http.response.Response;
import scala.Function1;
import scala.Function2;
import scala.PartialFunction;
import scala.Tuple2;

import javax.net.ssl.KeyManagerFactory;
import java.net.InetSocketAddress;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;

import static io.gatling.core.javaapi.internal.ScalaHelpers.*;
import static io.gatling.http.javaapi.internal.HttpChecks.*;

public final class HttpProtocolBuilder implements ProtocolBuilder {
  private final io.gatling.http.protocol.HttpProtocolBuilder wrapped;

  public HttpProtocolBuilder(io.gatling.http.protocol.HttpProtocolBuilder wrapped) {
    this.wrapped = wrapped;
  }

  public io.gatling.core.protocol.Protocol protocol() {
    return wrapped.protocol();
  }

  public HttpProtocolBuilder baseUrl(String url) {
    return baseUrls(url);
  }

  public HttpProtocolBuilder baseUrls(String... urls) {
    return baseUrls(Arrays.asList(urls));
  }

  public HttpProtocolBuilder baseUrls(List<String> urls) {
    return new HttpProtocolBuilder(wrapped.baseUrls(toScalaSeq(urls)));
  }

  public HttpProtocolBuilder warmUp(String url) {
    return new HttpProtocolBuilder(wrapped.warmUp(url));
  }

  public HttpProtocolBuilder disableWarmUp() {
    return new HttpProtocolBuilder(wrapped.disableWarmUp());
  }

  // enginePart
  public HttpProtocolBuilder shareConnections() {
    return new HttpProtocolBuilder(wrapped.shareConnections());
  }

  public HttpProtocolBuilder virtualHost(String virtualHost) {
    return new HttpProtocolBuilder(wrapped.virtualHost(toStringExpression(virtualHost)));
  }

  public HttpProtocolBuilder virtualHost(Function<Session, String> virtualHost) {
    return new HttpProtocolBuilder(wrapped.virtualHost(toTypedGatlingSessionFunction(virtualHost)));
  }

  public HttpProtocolBuilder localAddress(String address) {
    return new HttpProtocolBuilder(wrapped.localAddress(address));
  }

  public HttpProtocolBuilder localAddresses(String... addresses) {
    return new HttpProtocolBuilder(wrapped.localAddresses(toScalaSeq(addresses)));
  }

  public HttpProtocolBuilder localAddresses(List<String> addresses) {
    return new HttpProtocolBuilder(wrapped.localAddresses(toScalaSeq(addresses)));
  }

  public HttpProtocolBuilder useAllLocalAddresses() {
    return new HttpProtocolBuilder(wrapped.useAllLocalAddresses());
  }

  public HttpProtocolBuilder useAllLocalAddressesMatching(String... patterns) {
    return new HttpProtocolBuilder(wrapped.useAllLocalAddressesMatching(toScalaSeq(patterns)));
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeFirefoxOld() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeFirefoxOld());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeFirefox() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeFirefox());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeOperaOld() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeOperaOld());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeOpera() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeOpera());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeSafariOld() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeSafariOld());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeSafari() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeSafari());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeIE7() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeIE7());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeIE8() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeIE8());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeIE10() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeIE10());
  }

  public HttpProtocolBuilder maxConnectionsPerHostLikeChrome() {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHostLikeChrome());
  }

  public HttpProtocolBuilder maxConnectionsPerHost(int max) {
    return new HttpProtocolBuilder(wrapped.maxConnectionsPerHost(max));
  }

  public HttpProtocolBuilder perUserKeyManagerFactory(Function<Long, KeyManagerFactory> f) {
    return new HttpProtocolBuilder(wrapped.perUserKeyManagerFactory(untyped -> f.apply((Long) untyped)));
  }

  // requestPart
  public HttpProtocolBuilder disableAutoReferer() {
    return new HttpProtocolBuilder(wrapped.disableAutoReferer());
  }

  public HttpProtocolBuilder disableAutoOrigin() {
    return new HttpProtocolBuilder(wrapped.disableAutoOrigin());
  }

  public HttpProtocolBuilder disableCaching() {
    return new HttpProtocolBuilder(wrapped.disableCaching());
  }

  public HttpProtocolBuilder header(CharSequence name, String value) {
    return new HttpProtocolBuilder(wrapped.header(name, toStringExpression(value)));
  }

  public HttpProtocolBuilder header(CharSequence name, Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.header(name, toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder headers(Map<? extends CharSequence, String> headers) {
    return new HttpProtocolBuilder(wrapped.headers(toScalaMap(headers)));
  }

  public HttpProtocolBuilder acceptHeader(String value) {
    return new HttpProtocolBuilder(wrapped.acceptHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder acceptHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.acceptHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder acceptCharsetHeader(String value) {
    return new HttpProtocolBuilder(wrapped.acceptCharsetHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder acceptCharsetHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.acceptCharsetHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder acceptEncodingHeader(String value) {
    return new HttpProtocolBuilder(wrapped.acceptEncodingHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder acceptEncodingHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.acceptEncodingHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder acceptLanguageHeader(String value) {
    return new HttpProtocolBuilder(wrapped.acceptLanguageHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder acceptLanguageHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.acceptLanguageHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder authorizationHeader(String value) {
    return new HttpProtocolBuilder(wrapped.authorizationHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder authorizationHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.authorizationHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder connectionHeader(String value) {
    return new HttpProtocolBuilder(wrapped.connectionHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder connectionHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.connectionHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder contentTypeHeader(String value) {
    return new HttpProtocolBuilder(wrapped.contentTypeHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder contentTypeHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.contentTypeHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder doNotTrackHeader(String value) {
    return new HttpProtocolBuilder(wrapped.doNotTrackHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder doNotTrackHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.doNotTrackHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder originHeader(String value) {
    return new HttpProtocolBuilder(wrapped.originHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder originHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.originHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder userAgentHeader(String value) {
    return new HttpProtocolBuilder(wrapped.userAgentHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder userAgentHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.userAgentHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder upgradeInsecureRequestsHeader(String value) {
    return new HttpProtocolBuilder(wrapped.upgradeInsecureRequestsHeader(toStringExpression(value)));
  }

  public HttpProtocolBuilder upgradeInsecureRequestsHeader(Function<Session, String> value) {
    return new HttpProtocolBuilder(wrapped.upgradeInsecureRequestsHeader(toTypedGatlingSessionFunction(value)));
  }

  public HttpProtocolBuilder basicAuth(String username, String password) {
    return new HttpProtocolBuilder(wrapped.basicAuth(toStringExpression(username), toStringExpression(password)));
  }

  public HttpProtocolBuilder basicAuth(String username, Function<Session, String> password) {
    return new HttpProtocolBuilder(wrapped.basicAuth(toStringExpression(username), toTypedGatlingSessionFunction(password)));
  }

  public HttpProtocolBuilder basicAuth(Function<Session, String> username, String password) {
    return new HttpProtocolBuilder(wrapped.basicAuth(toTypedGatlingSessionFunction(username), toStringExpression(password)));
  }

  public HttpProtocolBuilder basicAuth(Function<Session, String> username, Function<Session, String> password) {
    return new HttpProtocolBuilder(wrapped.basicAuth(toTypedGatlingSessionFunction(username), toTypedGatlingSessionFunction(password)));
  }

  public HttpProtocolBuilder digestAuth(String username, String password) {
    return new HttpProtocolBuilder(wrapped.digestAuth(toStringExpression(username), toStringExpression(password)));
  }

  public HttpProtocolBuilder digestAuth(String username, Function<Session, String> password) {
    return new HttpProtocolBuilder(wrapped.digestAuth(toStringExpression(username), toTypedGatlingSessionFunction(password)));
  }

  public HttpProtocolBuilder digestAuth(Function<Session, String> username, String password) {
    return new HttpProtocolBuilder(wrapped.digestAuth(toTypedGatlingSessionFunction(username), toStringExpression(password)));
  }

  public HttpProtocolBuilder digestAuth(Function<Session, String> username, Function<Session, String> password) {
    return new HttpProtocolBuilder(wrapped.digestAuth(toTypedGatlingSessionFunction(username), toTypedGatlingSessionFunction(password)));
  }

  public HttpProtocolBuilder silentResources() {
    return new HttpProtocolBuilder(wrapped.silentResources());
  }

  public HttpProtocolBuilder silentUri(String regex) {
    return new HttpProtocolBuilder(wrapped.silentUri(regex));
  }

  public HttpProtocolBuilder disableUrlEncoding() {
    return new HttpProtocolBuilder(wrapped.disableUrlEncoding());
  }

  public HttpProtocolBuilder sign(SignatureCalculator calculator) {
    return new HttpProtocolBuilder(wrapped.sign(toStaticValueExpression(calculator)));
  }

  public HttpProtocolBuilder sign(Function<Session, SignatureCalculator> calculator) {
    return new HttpProtocolBuilder(wrapped.sign(toTypedGatlingSessionFunction(calculator)));
  }

  public HttpProtocolBuilder signWithOAuth1(String consumerKey, String clientSharedSecret, String token, String tokenSecret) {
    return new HttpProtocolBuilder(wrapped.signWithOAuth1(toStringExpression(consumerKey), toStringExpression(clientSharedSecret), toStringExpression(token), toStringExpression(tokenSecret)));
  }

  public HttpProtocolBuilder signWithOAuth1(Function<Session, String> consumerKey, Function<Session, String> clientSharedSecret, Function<Session, String> token, Function<Session, String> tokenSecret) {
    return new HttpProtocolBuilder(wrapped.signWithOAuth1(toTypedGatlingSessionFunction(consumerKey), toTypedGatlingSessionFunction(clientSharedSecret), toTypedGatlingSessionFunction(token), toTypedGatlingSessionFunction(tokenSecret)));
  }

  public HttpProtocolBuilder enableHttp2() {
    return new HttpProtocolBuilder(wrapped.enableHttp2());
  }

  public HttpProtocolBuilder http2PriorKnowledge(Map<String, Boolean> remotes) {
    return new HttpProtocolBuilder(wrapped.http2PriorKnowledge(toScalaMap(remotes.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)))));
  }

  // responsePart
  public HttpProtocolBuilder disableFollowRedirect() {
    return new HttpProtocolBuilder(wrapped.disableFollowRedirect());
  }

  public HttpProtocolBuilder maxRedirects(int max) {
    return new HttpProtocolBuilder(wrapped.maxRedirects(max));
  }

  public HttpProtocolBuilder strict302Handling() {
    return new HttpProtocolBuilder(wrapped.strict302Handling());
  }

  public HttpProtocolBuilder transformResponse(BiFunction<Response, Session, Response> f) {
    return new HttpProtocolBuilder(wrapped.transformResponse(toTypedGatlingSessionFunction(f)));
  }

  public HttpProtocolBuilder check(CheckBuilder... checks) {
    return check(Arrays.asList(checks));
  }

  public HttpProtocolBuilder check(List<CheckBuilder> checks) {
    return new HttpProtocolBuilder(wrapped.check(toScalaChecks(checks)));
  }

  public UntypedCondition checkIf(Function<Session, Boolean> condition) {
    return new UntypedCondition(toUntypedGatlingSessionFunction(condition));
  }

  public UntypedCondition checkIf(String condition) {
    return new UntypedCondition(toBooleanExpression(condition));
  }

  public final class UntypedCondition {
    private final Function1<io.gatling.core.session.Session, Validation<Object>> condition;

    public UntypedCondition(Function1<io.gatling.core.session.Session, Validation<Object>> condition) {
      this.condition = condition;
    }

    public HttpProtocolBuilder then(CheckBuilder... thenChecks) {
      return then(Arrays.asList(thenChecks));
    }

    public HttpProtocolBuilder then(List<CheckBuilder> thenChecks) {
      return new HttpProtocolBuilder(wrapped.checkIf(condition, toScalaChecks(thenChecks)));
    }
  }

  public TypedCondition checkIf(BiFunction<Response, Session, Boolean> condition) {
    return new TypedCondition(toUntypedGatlingSessionFunction(condition));
  }

  public final class TypedCondition {
    private final Function2<Response, io.gatling.core.session.Session, Validation<Object>> condition;

    public TypedCondition(Function2<Response, io.gatling.core.session.Session, Validation<Object>> condition) {
      this.condition = condition;
    }

    public HttpProtocolBuilder then(CheckBuilder... thenChecks) {
      return then(Arrays.asList(thenChecks));
    }

    public HttpProtocolBuilder then(List<CheckBuilder> thenChecks) {
      return new HttpProtocolBuilder(wrapped.checkIf(condition, toScalaChecks(thenChecks)));
    }
  }

  public HttpProtocolBuilder inferHtmlResources() {
    return new HttpProtocolBuilder(wrapped.inferHtmlResources());
  }

  public HttpProtocolBuilder inferHtmlResources(Filter.AllowList allow) {
    return new HttpProtocolBuilder(wrapped.inferHtmlResources(allow.asScala()));
  }

  public HttpProtocolBuilder inferHtmlResources(Filter.AllowList allow, Filter.DenyList deny) {
    return new HttpProtocolBuilder(wrapped.inferHtmlResources(allow.asScala(), deny.asScala()));
  }

  public HttpProtocolBuilder inferHtmlResources(Filter.DenyList deny) {
    return new HttpProtocolBuilder(wrapped.inferHtmlResources(deny.asScala()));
  }

  public HttpProtocolBuilder inferHtmlResources(Filter.DenyList deny, Filter.AllowList allow) {
    return new HttpProtocolBuilder(wrapped.inferHtmlResources(deny.asScala(), allow.asScala()));
  }

  public HttpProtocolBuilder nameInferredHtmlResourcesAfterUrlTail() {
    return new HttpProtocolBuilder(wrapped.nameInferredHtmlResourcesAfterUrlTail());
  }

  public HttpProtocolBuilder nameInferredHtmlResourcesAfterAbsoluteUrl() {
    return new HttpProtocolBuilder(wrapped.nameInferredHtmlResourcesAfterAbsoluteUrl());
  }

  public HttpProtocolBuilder nameInferredHtmlResourcesAfterRelativeUrl() {
    return new HttpProtocolBuilder(wrapped.nameInferredHtmlResourcesAfterRelativeUrl());
  }

  public HttpProtocolBuilder nameInferredHtmlResourcesAfterPath() {
    return new HttpProtocolBuilder(wrapped.nameInferredHtmlResourcesAfterPath());
  }

  public HttpProtocolBuilder nameInferredHtmlResourcesAfterLastPathElement() {
    return new HttpProtocolBuilder(wrapped.nameInferredHtmlResourcesAfterLastPathElement());
  }

  public HttpProtocolBuilder nameInferredHtmlResources(Function<Uri, String> f) {
    return new HttpProtocolBuilder(wrapped.nameInferredHtmlResources(f::apply));
  }

  // wsPart
  public HttpProtocolBuilder wsBaseUrl(String url) {
    return new HttpProtocolBuilder(wrapped.wsBaseUrl(url));
  }

  public HttpProtocolBuilder wsBaseUrls(String... urls) {
    return new HttpProtocolBuilder(wrapped.wsBaseUrls(toScalaSeq(urls)));
  }

  public HttpProtocolBuilder wsBaseUrls(List<String> urls) {
    return new HttpProtocolBuilder(wrapped.wsBaseUrls(toScalaSeq(urls)));
  }

  public HttpProtocolBuilder wsReconnect() {
    return new HttpProtocolBuilder(wrapped.wsReconnect());
  }

  public HttpProtocolBuilder wsMaxReconnects(int max) {
    return new HttpProtocolBuilder(wrapped.wsMaxReconnects(max));
  }

  public HttpProtocolBuilder wsAutoReplyTextFrame(Function<String, String> f) {
    return new HttpProtocolBuilder(wrapped.wsAutoReplyTextFrame(new PartialFunction<String, String>() {
      @Override
      public boolean isDefinedAt(String x) {
        return f.apply(x) != null;
      }

      @Override
      public String apply(String v1) {
        return f.apply(v1);
      }
    }));
  }

  public HttpProtocolBuilder wsAutoReplySocketIo4() {
    return new HttpProtocolBuilder(wrapped.wsAutoReplySocketIo4());
  }

  // proxyPart
  public HttpProtocolBuilder noProxyFor(String... hosts) {
    return new HttpProtocolBuilder(wrapped.noProxyFor(toScalaSeq(hosts)));
  }

  public HttpProtocolBuilder proxy(Proxy proxy) {
    return new HttpProtocolBuilder(wrapped.proxy(proxy.asScala()));
  }

  // dnsPart
  public HttpProtocolBuilder asyncNameResolution(String... dnsServers) {
    return new HttpProtocolBuilder(wrapped.asyncNameResolution(toScalaSeq(dnsServers)));
  }

  public HttpProtocolBuilder asyncNameResolution(InetSocketAddress[] dnsServers) {
    return new HttpProtocolBuilder(wrapped.asyncNameResolution(dnsServers));
  }

  public HttpProtocolBuilder perUserNameResolution() {
    return new HttpProtocolBuilder(wrapped.perUserNameResolution());
  }
}
